Missing table 'anlageneeggeothermiegrubengasdruckentspannung'

## Description of the issue
It seems like new tables were introduced, so we have to adapt the `orm.py` file.

## Steps to Reproduce
1. Run the full bulk download.
2. Get the Error: `KeyError: 'anlageneeggeothermiegrubengasdruckentspannung'`

## Context and Environment
* Version used: 0.13.0
* Operating system: Windows
* Environment setup and (python) version: 3.10

## Workflow checklist
- [x] I am aware of the workflow in [CONTRIBUTING.md](https://github.com/OpenEnergyPlatform/open-MaStR/blob/production/CONTRIBUTING.md)

@chrwm I get the following error when running the tests:

FAILED tests/soap_api/test_mastr_mirror.py::test_retrieve_additional_data - TypeError: 'Netzbetreiberzuordnungen' is an invalid keyword argument for StorageExtended

--- a/.bumpversion.cfg
+++ b/.bumpversion.cfg
@@ -1,5 +1,5 @@
 [bumpversion]
-current_version = 0.13.0
+current_version = 0.13.1
 parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)((?P<release>(a|na))+(?P<build>\d+))?
 serialize = 
 	{major}.{minor}.{patch}{release}{build}

--- a/.github/workflows/ci-production.yml
+++ b/.github/workflows/ci-production.yml
@@ -34,7 +34,7 @@ jobs:
     - name: create package
       run: python setup.py sdist
     - name: import open-mastr
-      run: python -m pip install ./dist/open_mastr-0.13.0.tar.gz
+      run: python -m pip install ./dist/open_mastr-0.13.1.tar.gz
     - name: Create credentials file
       env:
         MASTR_TOKEN: ${{ secrets.MASTR_TOKEN }}

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,7 @@ and the versioning aims to respect [Semantic Versioning](http://semver.org/spec/
 ## [v0.13.1] Hotfix - 2023-04-11
 
 ### Added
-- Add new table and new columns to the data model [#439](https://github.com/OpenEnergyPlatform/open-MaStR/pull/439)
+- Add new table and new columns to the data model [#440](https://github.com/OpenEnergyPlatform/open-MaStR/pull/440)
 ## [v0.13.0] Maintenance release - 2023-02-16
 ### Added
 - Add a `workflow_dispatch` to run CI pipelines from a button click [#389](https://github.com/OpenEnergyPlatform/open-MaStR/pull/389)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,10 @@ For each version important additions, changes and removals are listed here.
 The format is inspired from [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and the versioning aims to respect [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [v0.13.1] Hotfix - 2023-04-11
 
+### Added
+- Add new table and new columns to the data model [#439](https://github.com/OpenEnergyPlatform/open-MaStR/pull/439)
 ## [v0.13.0] Maintenance release - 2023-02-16
 ### Added
 - Add a `workflow_dispatch` to run CI pipelines from a button click [#389](https://github.com/OpenEnergyPlatform/open-MaStR/pull/389)

--- a/CITATION.cff
+++ b/CITATION.cff
@@ -28,7 +28,7 @@ authors:
 title: "open-MaStR"
 type: software
 license: AGPL-3.0
-version: 0.13.0
+version: 0.13.1
 doi: 
-date-released: 2023-02-16
+date-released: 2023-04-11
 url: "https://github.com/OpenEnergyPlatform/open-MaStR/"

--- a/open_mastr/mastr.py
+++ b/open_mastr/mastr.py
@@ -134,6 +134,7 @@ class Mastr:
                 "balancing_area", "Yes", "No"
                 "permit", "Yes", "Yes"
                 "deleted_units", "Yes", "No"
+                "retrofit_units", "Yes", "No"
 
         date: None or :class:`datetime.datetime` or str, optional
             For bulk method:

--- a/open_mastr/soap_api/download.py
+++ b/open_mastr/soap_api/download.py
@@ -479,7 +479,7 @@ class MaStRDownload:
                 "permit_data": "GetEinheitGenehmigung",
             },
             "gsgk": {
-                "unit_data": "GetEinheitGeoSolarthermieGrubenKlaerschlammDruckentspannung",
+                "unit_data": "GetEinheitGeothermieGrubengasDruckentspannung",
                 "energietraeger": [
                     "Geothermie",
                     "Solarthermie",

--- a/open_mastr/soap_api/download.py
+++ b/open_mastr/soap_api/download.py
@@ -487,7 +487,7 @@ class MaStRDownload:
                     "Klaerschlamm",
                 ],
                 "kwk_data": "GetAnlageKwk",
-                "eeg_data": "GetAnlageEegGeoSolarthermieGrubenKlaerschlammDruckentspannung",
+                "eeg_data": "GetAnlageEegGeothermieGrubengasDruckentspannung",
                 "permit_data": "GetEinheitGenehmigung",
             },
             "nuclear": {

--- a/open_mastr/soap_api/mirror.py
+++ b/open_mastr/soap_api/mirror.py
@@ -318,7 +318,7 @@ class MaStRMirror:
                     data, requested_ids, download_functions[data_type]
                 )
 
-                unit_data = flatten_dict(unit_data)
+                unit_data = flatten_dict(unit_data, serialize_with_json=False)
                 number_units_merged = 0
 
                 # Prepare data and add to database table
@@ -839,20 +839,21 @@ class MaStRMirror:
                 ertuechtigung["ProzentualeErhoehungDesLv"] = float(
                     ertuechtigung["ProzentualeErhoehungDesLv"]
                 )
-        # The NetzbetreiberMastrNummer is handed over as type:list, hence
-        # non-compatible with sqlite)
-        # This replaces the list with the first (string)element in the list
-        # to make it sqlite compatible
-        if (
-            "NetzbetreiberMastrNummer" in unit_dat
-            and type(unit_dat["NetzbetreiberMastrNummer"]) == list
-        ):
-            if len(unit_dat["NetzbetreiberMastrNummer"]) > 0:
-                unit_dat["NetzbetreiberMastrNummer"] = unit_dat[
-                    "NetzbetreiberMastrNummer"
-                ][0]
-            else:
-                unit_dat["NetzbetreiberMastrNummer"] = None
+        # Some data (data_in_list) is handed over as type:list, hence
+        # non-compatible with sqlite or postgresql
+        # This replaces the list with the first element in the list
+
+        data_as_list = ["NetzbetreiberMastrNummer","Netzbetreiberzuordnungen"]
+
+        for dat in data_as_list:
+            if (
+                dat in unit_dat
+                and type(unit_dat[dat]) == list
+            ):
+                if len(unit_dat[dat]) > 0:
+                    unit_dat[dat] = f"{unit_dat[dat][0]}"
+                else:
+                    unit_dat[dat] = None
 
         # Rename the typo in column zugeordneteWirkleistungWechselrichter
         if "zugeordneteWirkleistungWechselrichter" in unit_dat.keys():

--- a/open_mastr/utils/constants.py
+++ b/open_mastr/utils/constants.py
@@ -16,6 +16,7 @@ BULK_DATA = [
     "balancing_area",
     "permit",
     "deleted_units",
+    "retrofit_units",
 ]
 
 # Possible values for parameter 'data' with API download method
@@ -99,6 +100,7 @@ BULK_INCLUDE_TABLES_MAP = {
     "balancing_area": ["bilanzierungsgebiete"],
     "permit": ["einheitengenehmigung"],
     "deleted_units": ["geloeschteunddeaktivierteeinheiten"],
+    "retrofit_units": ["ertuechtigungen"],
 }
 
 # Map bulk data to database table names, for csv export

--- a/open_mastr/utils/constants.py
+++ b/open_mastr/utils/constants.py
@@ -16,7 +16,6 @@ BULK_DATA = [
     "balancing_area",
     "permit",
     "deleted_units",
-    "retrofit_units",
 ]
 
 # Possible values for parameter 'data' with API download method
@@ -100,7 +99,6 @@ BULK_INCLUDE_TABLES_MAP = {
     "balancing_area": ["bilanzierungsgebiete"],
     "permit": ["einheitengenehmigung"],
     "deleted_units": ["geloeschteunddeaktivierteeinheiten"],
-    "retrofit_units": ["ertuechtigungen"],
 }
 
 # Map bulk data to database table names, for csv export

--- a/open_mastr/utils/constants.py
+++ b/open_mastr/utils/constants.py
@@ -16,6 +16,7 @@ BULK_DATA = [
     "balancing_area",
     "permit",
     "deleted_units",
+    "retrofit_units",
 ]
 
 # Possible values for parameter 'data' with API download method
@@ -59,7 +60,7 @@ ADDITIONAL_TABLES = [
     "market_actors",
     "market_roles",
     "permit",
-    "deleted_units"
+    "deleted_units",
 ]
 
 # Possible data types for API download
@@ -99,6 +100,7 @@ BULK_INCLUDE_TABLES_MAP = {
     "balancing_area": ["bilanzierungsgebiete"],
     "permit": ["einheitengenehmigung"],
     "deleted_units": ["geloeschteunddeaktivierteeinheiten"],
+    "retrofit_units": ["ertuechtigungen"],
 }
 
 # Map bulk data to database table names, for csv export
@@ -152,10 +154,7 @@ ORM_MAP = {
         "eeg_data": "HydroEeg",
         "permit_data": "Permit",
     },
-    "nuclear": {
-        "unit_data": "NuclearExtended",
-        "permit_data": "Permit"
-    },
+    "nuclear": {"unit_data": "NuclearExtended", "permit_data": "Permit"},
     "storage": {
         "unit_data": "StorageExtended",
         "eeg_data": "StorageEeg",
@@ -173,9 +172,8 @@ ORM_MAP = {
     "grids": "Grids",
     "balancing_area": "BalancingArea",
     "permit": "Permit",
-    "deleted_units": "DeletedUnits"
+    "deleted_units": "DeletedUnits",
 }
-
 
 
 UNIT_TYPE_MAP = {

--- a/open_mastr/utils/constants.py
+++ b/open_mastr/utils/constants.py
@@ -61,6 +61,7 @@ ADDITIONAL_TABLES = [
     "market_roles",
     "permit",
     "deleted_units",
+    "retrofit_units",
 ]
 
 # Possible data types for API download
@@ -118,6 +119,7 @@ BULK_ADDITIONAL_TABLES_CSV_EXPORT_MAP = {
     "balancing_area": ["balancing_area"],
     "permit": ["permit"],
     "deleted_units": ["deleted_units"],
+    "retrofit_units": ["retrofit_units"],
 }
 
 # used to map the parameter options in open-mastr to the exact table class names in orm.py
@@ -173,6 +175,7 @@ ORM_MAP = {
     "balancing_area": "BalancingArea",
     "permit": "Permit",
     "deleted_units": "DeletedUnits",
+    "retrofit_units": "RetrofitUnits",
 }
 
 

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -228,6 +228,7 @@ class CombustionExtended(Extended, ParentAllTables, Base):
     Einsatzort = Column(String)
     KwkMastrNummer = Column(String)
     Technologie = Column(String)
+    AusschliesslicheVerwendungImKombibetrieb = Column(Boolean)
 
 
 class GsgkExtended(Extended, ParentAllTables, Base):

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -135,6 +135,7 @@ class Extended(object):
     Einspeisungsart = Column(String)
     PraequalifiziertFuerRegelenergie = Column(Boolean)
     GenMastrNummer = Column(String)
+    Netzbetreiberzuordnungen = Column(String)
     # from bulk download
     Hausnummer_nv = Column(Boolean)
     Weic_nv = Column(Boolean)
@@ -294,6 +295,7 @@ class Eeg(object):
     AusschreibungZuschlag = Column(Boolean)
     AnlagenkennzifferAnlagenregister = Column(String)
     AnlagenkennzifferAnlagenregister_nv = Column(Boolean)
+    Netzbetreiberzuordnungen = Column(String)
 
 
 class WindEeg(Eeg, ParentAllTables, Base):
@@ -387,6 +389,7 @@ class Kwk(ParentAllTables, Base):
     VerknuepfteEinheiten = Column(String)
     AnlageBetriebsstatus = Column(String)
     AusschreibungZuschlag = Column(Boolean)
+    Netzbetreiberzuordnungen = Column(String)
 
 
 class Permit(ParentAllTables, Base):
@@ -406,6 +409,7 @@ class Permit(ParentAllTables, Base):
     VerknuepfteEinheiten = Column(String)
     Frist_nv = Column(Boolean)
     WasserrechtAblaufdatum_nv = Column(Boolean)
+    Netzbetreiberzuordnungen = Column(String)
 
 
 class LocationBasic(Base):

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -741,6 +741,7 @@ class Grids(ParentAllTables, Base):
     GeschlossenesVerteilnetz = Column(String)
     Bezeichnung = Column(String)
     Marktgebiet = Column(String)
+    Bundesland = Column(String)
 
 
 class GridConnections(ParentAllTables, Base):
@@ -774,6 +775,18 @@ class DeletedUnits(ParentAllTables, Base):
     EinheitBetriebsstatus = Column(String)
 
 
+class RetrofitUnits(ParentAllTables, Base):
+    __tablename__ = "retrofit_units"
+
+    Id = Column(Integer, primary_key=True)
+    EegMastrNummer = Column(String)
+    Leistungserhoehung = Column(Float)
+    WiederinbetriebnahmeDatum = Column(Date)
+    DatumLetzteAktualisierung = Column(DateTime(timezone=True))
+    Ertuechtigungsart = Column(String)
+    ErtuechtigungIstZulassungspflichtig = Column(Boolean)
+
+
 tablename_mapping = {
     "anlageneegbiomasse": {
         "__name__": BiomassEeg.__tablename__,
@@ -793,7 +806,7 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
         },
     },
-    "anlageneeggeosolarthermiegrubenklaerschlammdruckentspannung": {
+    "anlageneeggeothermiegrubengasdruckentspannung": {
         "__name__": GsgkEeg.__tablename__,
         "__class__": GsgkEeg,
         "replace_column_names": {
@@ -801,7 +814,7 @@ tablename_mapping = {
             "VerknuepfteEinheitenMaStRNummern": "VerknuepfteEinheit",
         },
     },
-    "einheitengeosolarthermiegrubenklaerschlammdruckentspannung": {
+    "einheitengeothermiegrubengasdruckentspannung": {
         "__name__": GsgkExtended.__tablename__,
         "__class__": GsgkExtended,
         "replace_column_names": {
@@ -957,6 +970,11 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
             "KwkMaStRNummer": "KwkMastrNummer",
         },
+    },
+    "ertuechtigungen": {
+        "__name__": RetrofitUnits.__tablename__,
+        "__class__": RetrofitUnits,
+        "replace_column_names": None,
     },
     "geloeschteunddeaktivierteeinheiten": {
         "__name__": DeletedUnits.__tablename__,

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -776,6 +776,18 @@ class DeletedUnits(ParentAllTables, Base):
     EinheitBetriebsstatus = Column(String)
 
 
+class RetrofitUnits(ParentAllTables, Base):
+    __tablename__ = "retrofit_units"
+
+    Id = Column(Integer, primary_key=True)
+    EegMastrNummer = Column(String)
+    Leistungserhoehung = Column(Float)
+    WiederinbetriebnahmeDatum = Column(Date)
+    DatumLetzteAktualisierung = Column(DateTime(timezone=True))
+    Ertuechtigungsart = Column(String)
+    ErtuechtigungIstZulassungspflichtig = Column(Boolean)
+
+
 tablename_mapping = {
     "anlageneegbiomasse": {
         "__name__": BiomassEeg.__tablename__,
@@ -795,7 +807,7 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
         },
     },
-    "anlageneeggeosolarthermiegrubenklaerschlammdruckentspannung": {
+    "anlageneeggeothermiegrubengasdruckentspannung": {
         "__name__": GsgkEeg.__tablename__,
         "__class__": GsgkEeg,
         "replace_column_names": {
@@ -803,7 +815,7 @@ tablename_mapping = {
             "VerknuepfteEinheitenMaStRNummern": "VerknuepfteEinheit",
         },
     },
-    "einheitengeosolarthermiegrubenklaerschlammdruckentspannung": {
+    "einheitengeothermiegrubengasdruckentspannung": {
         "__name__": GsgkExtended.__tablename__,
         "__class__": GsgkExtended,
         "replace_column_names": {
@@ -959,6 +971,11 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
             "KwkMaStRNummer": "KwkMastrNummer",
         },
+    },
+    "ertuechtigungen": {
+        "__name__": RetrofitUnits.__tablename__,
+        "__class__": RetrofitUnits,
+        "replace_column_names": None,
     },
     "geloeschteunddeaktivierteeinheiten": {
         "__name__": DeletedUnits.__tablename__,

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -42,6 +42,7 @@ class BasicUnit(Base):
     GenMastrNummer = Column(String)
     BestandsanlageMastrNummer = Column(String)
     NichtVorhandenInMigriertenEinheiten = Column(String)
+    EinheitSystemstatus = Column(String)
 
 
 class AdditionalDataRequested(Base):

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -743,6 +743,7 @@ class Grids(ParentAllTables, Base):
     GeschlossenesVerteilnetz = Column(String)
     Bezeichnung = Column(String)
     Marktgebiet = Column(String)
+    Bundesland = Column(String)
 
 
 class GridConnections(ParentAllTables, Base):

--- a/open_mastr/utils/orm.py
+++ b/open_mastr/utils/orm.py
@@ -741,7 +741,6 @@ class Grids(ParentAllTables, Base):
     GeschlossenesVerteilnetz = Column(String)
     Bezeichnung = Column(String)
     Marktgebiet = Column(String)
-    Bundesland = Column(String)
 
 
 class GridConnections(ParentAllTables, Base):
@@ -775,18 +774,6 @@ class DeletedUnits(ParentAllTables, Base):
     EinheitBetriebsstatus = Column(String)
 
 
-class RetrofitUnits(ParentAllTables, Base):
-    __tablename__ = "retrofit_units"
-
-    Id = Column(Integer, primary_key=True)
-    EegMastrNummer = Column(String)
-    Leistungserhoehung = Column(Float)
-    WiederinbetriebnahmeDatum = Column(Date)
-    DatumLetzteAktualisierung = Column(DateTime(timezone=True))
-    Ertuechtigungsart = Column(String)
-    ErtuechtigungIstZulassungspflichtig = Column(Boolean)
-
-
 tablename_mapping = {
     "anlageneegbiomasse": {
         "__name__": BiomassEeg.__tablename__,
@@ -806,7 +793,7 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
         },
     },
-    "anlageneeggeothermiegrubengasdruckentspannung": {
+    "anlageneeggeosolarthermiegrubenklaerschlammdruckentspannung": {
         "__name__": GsgkEeg.__tablename__,
         "__class__": GsgkEeg,
         "replace_column_names": {
@@ -814,7 +801,7 @@ tablename_mapping = {
             "VerknuepfteEinheitenMaStRNummern": "VerknuepfteEinheit",
         },
     },
-    "einheitengeothermiegrubengasdruckentspannung": {
+    "einheitengeosolarthermiegrubenklaerschlammdruckentspannung": {
         "__name__": GsgkExtended.__tablename__,
         "__class__": GsgkExtended,
         "replace_column_names": {
@@ -970,11 +957,6 @@ tablename_mapping = {
             "LokationMaStRNummer": "LokationMastrNummer",
             "KwkMaStRNummer": "KwkMastrNummer",
         },
-    },
-    "ertuechtigungen": {
-        "__name__": RetrofitUnits.__tablename__,
-        "__class__": RetrofitUnits,
-        "replace_column_names": None,
     },
     "geloeschteunddeaktivierteeinheiten": {
         "__name__": DeletedUnits.__tablename__,

--- a/open_mastr/xml_download/utils_cleansing_bulk.py
+++ b/open_mastr/xml_download/utils_cleansing_bulk.py
@@ -34,7 +34,6 @@ def replace_mastr_katalogeintraege(
     """Replaces the IDs from the mastr database by its mapped string values from
     the table katalogwerte"""
     katalogwerte = create_katalogwerte_from_bulk_download(zipped_xml_file_path)
-
     for column_name in df.columns:
         if column_name in columns_replace_list:
             if df[column_name].dtype == "O":
@@ -43,6 +42,7 @@ def replace_mastr_katalogeintraege(
                     df[column_name]
                     .str.split(",", expand=True)
                     .apply(lambda x: x.str.strip())
+                    .replace("", None)
                     .astype("Int64")
                     .applymap(katalogwerte.get)
                     .agg(lambda d: ",".join(i for i in d if isinstance(i, str)), axis=1)

--- a/open_mastr/xml_download/utils_write_to_database.py
+++ b/open_mastr/xml_download/utils_write_to_database.py
@@ -285,9 +285,16 @@ def write_single_entries_until_not_unique_comes_up(
             key_list = (
                 pd.read_sql(sql=select(primary_key), con=con).values.squeeze().tolist()
             )
-    df = df.set_index(primary_key.name)
+
     len_df_before = len(df)
-    df = df.drop(labels=key_list, errors="ignore")
+    df = df.drop_duplicates(
+        subset=[primary_key.name]
+    )  # drop all entries with duplicated primary keys in the dataframe
+    df = df.set_index(primary_key.name)
+
+    df = df.drop(
+        labels=key_list, errors="ignore"
+    )  # drop primary keys that already exist in the table
     df = df.reset_index()
     print(f"{len_df_before-len(df)} entries already existed in the database.")
 
@@ -378,5 +385,3 @@ def handle_xml_syntax_error(data: bytes, err: Error) -> pd.DataFrame:
     df = pd.read_xml(decoded_data)
     print("One invalid xml expression was deleted.")
     return df
-
-

--- a/setup.py
+++ b/setup.py
@@ -18,14 +18,14 @@ setup(
         "open_mastr.utils.config",
         "open_mastr.xml_download",
     ],
-    version="0.13.0",
+    version="0.13.1",
     description="A package that provides an interface for downloading and"
     " processing the data of the Marktstammdatenregister (MaStR)",
     long_description=long_description,
     long_description_content_type="text/x-rst",
     url="https://github.com/OpenEnergyPlatform/open-MaStR",
     download_url="https://github.com/OpenEnergyPlatform/open-MaStR/archive"
-    "/refs/tags/v0.13.0.tar.gz",
+    "/refs/tags/v0.13.1.tar.gz",
     author="Open Energy Family",
     author_email="datenzentrum@rl-institut.de",
     maintainer="Ludwig Hülk",

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -65,6 +65,8 @@ def parameter_dict_working_list():
             "grid",
             "balancing_area",
             "permit",
+            "deleted_units",
+            "retrofit_units",
             None,
             ["wind", "solar"],
         ],

